sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlPartitioningSuite.scala

```diff
@@ -36,9 +35,31 @@ final class XmlPartitioningSuite extends SparkFunSuite with Matchers with Before
     try {
       val fileName = s"test-data/xml-resources/fias_house${if (large) ".large" else ""}.xml$suffix"
       val xmlFile = getClass.getClassLoader.getResource(fileName).getFile
-      val results = spark.read.option("rowTag", "House").option("mode", "FAILFAST").xml(xmlFile)
-      // Test file has 37 records; large file is 20x the records
-      assert(results.count() === (if (large) 740 else 37))
+      if (large) {
+        // The large file is invalid because it concatenates several XML files together and thus
+        // there are more one root tags, and each one has a BOM character at the beginning.
+
+        // In FAILFAST mode, we should throw an exception
+        val error = intercept[SparkException] {
+          spark.read.option("rowTag", "House").option("mode", "FAILFAST").xml(xmlFile)
+        }
+        checkError(
+          exception = error,
+          condition = "MALFORMED_RECORD_IN_PARSING.WITHOUT_SUGGESTION",
+          parameters = Map("badRecord" -> "_corrupt_record", "failFastMode" -> "FAILFAST")
+        )
+
+        // In PERMISSIVE mode, we should read the records in the first root tag and ignore the rest
+        // of the content
```

Comment:
what was the behavior without this optimization? We should call out all the behavior changes clearly.

Reply:
For an XML file with multiple root tags, all records across the root tags can be read properly before the changes. After the changes, only the records in the first root tag can be read. Will put this in the PR description.